crates/cargo-bsp/tests/integration_test.rs

```diff
@@ -76,3 +81,201 @@ fn exit_notif_without_shutdown() {
+        let req = bsp_server::Request {
+            id: TEST_REQUEST_ID.into(),
+            method: CargoFeaturesState::METHOD.into(),
+            params: Default::default(),
+        };
+        cl.send(&to_string(&req).unwrap());
+    }
+
+    fn send_enable_features_request(
+        cl: &mut Client,
+        package_id: &str,
+        features_to_enable: &[&str],
+    ) {
+        println!("Sending request: {:?}", features_to_enable);
+        let req = bsp_server::Request {
+            id: TEST_REQUEST_ID.into(),
+            method: EnableCargoFeatures::METHOD.into(),
+            params: serde_json::to_value(EnableCargoFeaturesParams {
+                package_id: package_id.to_string(),
+                features: features_to_enable.iter().map(|&s| s.into()).collect(),
+            })
+            .unwrap(),
+        };
+        cl.send(&to_string(&req).unwrap());
+    }
+
+    fn send_disable_features_request(
+        cl: &mut Client,
+        package_id: &str,
+        features_to_disable: &[&str],
+    ) {
+        let req = bsp_server::Request {
+            id: TEST_REQUEST_ID.into(),
+            method: DisableCargoFeatures::METHOD.into(),
+            params: serde_json::to_value(DisableCargoFeaturesParams {
+                package_id: package_id.to_string(),
+                features: features_to_disable.iter().map(|&s| s.into()).collect(),
+            })
+            .unwrap(),
+        };
+        cl.send(&to_string(&req).unwrap());
+    }
+
+    fn overwrite_cargo_toml_with_features(features_slice: &[(&str, &[&str])]) {
+        const TEST_PROJECT_AUTHOR: &str = "Test Author";
+        const TEST_PROJECT_VERSION: &str = "0.0.1";
+
+        let mut cargo_toml = CargoToml::builder();
+        cargo_toml
+            .name(TEST_PROJECT_NAME)
+            .version(TEST_PROJECT_VERSION)
+            .author(TEST_PROJECT_AUTHOR);
+
+        for &feature_with_deps in features_slice {
+            let mut f = TomlFeature::new(feature_with_deps.0);
+            for &dep in feature_with_deps.1 {
+                f.feature(dep);
+            }
+            cargo_toml.feature(f.build());
+        }
+
+        let cargo_toml_string = cargo_toml.build().unwrap().to_string();
+        File::create("Cargo.toml")
+            .unwrap()
+            .write_all(cargo_toml_string.as_bytes())
+            .unwrap();
+    }
+
+    // Function that creates new temporary project with Cargo.toml with features
+    // and sets it as current directory. The newly created directory has to be deleted.
+    fn create_mock_rust_project_with_features_and_set_it_as_current_dir(
+        features_slice: &[(&str, &[&str])],
+    ) {
+        Command::new(toolchain::cargo())
+            .args(["init", TEST_PROJECT_NAME])
+            .output()
+            .expect("Failed to create new temporary project for testing.");
+
+        std::env::set_current_dir(TEST_PROJECT_NAME).unwrap();
+
+        overwrite_cargo_toml_with_features(features_slice);
```

Comment:
a po co w ogóle wołać cargo init, skoro cargo.toml i tak jest potem zastąpione?

Reply:
i nie podoba mi się bardzo, że to się dzieje w podkatalogu katalogu ze źródłami

Reply:
A co do cargo init, to wołanie go jest łatwiejsze, bo nie trzeba wtedy tworzyć osobno folderu src i pliku main.ra